pkg/apis/internal.acorn.io/v1/memory.go

```diff
@@ -0,0 +1,123 @@
+)
+
+func ParseMemory(s []string) (Memory, error) {
+	result := Memory{}
+	for _, s := range s {
+		workload, memBytes, specific := strings.Cut(s, "=")
+
+		// If setting all, swap workload and memBytes
+		if !specific {
+			memBytes = workload
+			workload = ""
+		}
+
+		quantity, err := ParseQuantityString(memBytes)
+		if err != nil {
+			return Memory{}, err
+		}
+
+		result[workload] = &quantity
+	}
+	return result, nil
+}
+
+func ParseQuantityString(memory string) (int64, error) {
+	numInfo := literal.NumInfo{}
+	err := literal.ParseNum(memory, &numInfo)
+	if err != nil {
+		return -1, err
+	}
+
+	quantity, err := strconv.ParseInt(numInfo.String(), 10, 64)
+	if err != nil {
+		return -1, err
+	}
+
+	return quantity, nil
+}
+
+var (
+	ErrInvalidAcornMemory   = errors.New("invalid memory from Acornfile")
+	ErrInvalidSetMemory     = errors.New("invalid memory set by user")
+	ErrInvalidDefaultMemory = errors.New("invalid memory default")
+)
+
+// +k8s:deepcopy-gen=false
+type MemoryError struct {
+	Err      error
+	Quantity string
+}
+
+func MemoryToRequirements(specMem Memory, containerName string, container Container, specMemDefault, specMemMaximum *int64) (*corev1.ResourceRequirements, MemoryError) {
+	var memMaximum, memDefault int64
+	if specMemDefault != nil {
+		memDefault = *specMemDefault
+	}
+	if specMemMaximum != nil {
+		memMaximum = *specMemMaximum
+	}
+
+	// Determine which memory should be used to set the resource limit/requests. Gets set
+	// 4 ways: User setting a specific workload, user setting all workloads, Acornfile, or
+	// from the apiv1.Config default.
+	memBytes, err := memDefault, ErrInvalidDefaultMemory
+	if m, set := specMem[containerName]; set && m != nil {
+		err = ErrInvalidSetMemory
+		memBytes = *m
+	} else if specMem[""] != nil {
+		err = ErrInvalidSetMemory
+		memBytes = *specMem[""]
+	} else if container.Memory != nil {
+		err = ErrInvalidAcornMemory
+		memBytes = *container.Memory
+	}
+
+	// For maximum memory, 0 is equivalent to "unlimited"
+	if memMaximum != 0 {
+		var (
+			maxQuantity     = resource.NewQuantity(memMaximum, resource.BinarySI).String()
+			defaultQuantity = resource.NewQuantity(memDefault, resource.BinarySI).String()
+			bytesQuantity   = resource.NewQuantity(memBytes, resource.BinarySI).String()
+		)
+
+		if memBytes > memMaximum {
+			if memBytes == memDefault {
+				return nil, MemoryError{
+					Err: fmt.Errorf(
+						"%w: workload-memory-default set to %v but exceeds the workload-memory-maximum of %v",
+						err, defaultQuantity, maxQuantity),
+					Quantity: bytesQuantity}
+			}
+			return nil, MemoryError{
+				Err: fmt.Errorf(
+					"%w: workload %v with memory of %v exceeds the workload-memory-maximum of %v",
+					err, containerName, bytesQuantity, maxQuantity),
+				Quantity: bytesQuantity}
+		} else if memBytes == 0 {
+			// For bytes, 0 is viewed as the maximum allowed memory". As such,
+			// update to the current maximum.
+			memBytes = memMaximum
+		}
+	}
```

Comment:
I think this logic belongs in `validator.go`. The function should be responsible for parsing the memory information, but let the validator validate it.

Once the validator does its job, then any subsequent call to this method for parsing can safely assume the values meet the expected criteria, or that the app was in compliance when it was deployed and should continue to function properly.

Additionally, this would allow the app to continue to function as expected if the default memory for a cluster is updated and causes a workflow to be out of compliance. The offending app would only be flagged as invalid on update, which I think is the desired outcome.

Lastly, putting this code in the validator will also get rid of the need for the `MemoryError` type.

Reply:
Talked offline but the most recent push fixes this to be in the vision of what you were mentioning.